File: src/main/java/org/treblereel/gwt/three4g/math/Plane.java
package org.treblereel.gwt.three4g.math;

import jsinterop.annotations.JsConstructor;
import jsinterop.annotations.JsType;

/**
 * A two dimensional surface that extends infinitely in 3d space, represented in Hessian normal form by a unit
 * length normal vector and a constant.
 *
 * @author Dmitrii Tikhomirov <chani@me.com>
 * Created by treblereel on 3/1/18.
 */
@JsType(isNative = true, namespace = "THREE")
public class Plane {

    /**
     * a unit length Vector3 defining the normal of the plane. Default is (1, 0, 0).
     */
    public Vector3 normal;

    /**
     * the signed distance from the origin to the plane. Default is 0.
     */
    public float constant;

    @JsConstructor
    public Plane() {

    }

    @JsConstructor
    public Plane(Vector3 vector3, float constant) {

    }

    /**
     * Apply a Matrix4 to the plane. The matrix must be an affine, homogeneous transform.
     *
     * @param matrix - the Matrix4 to apply.ix
     * @return
     */

    public native Plane applyMatrix4(Matrix4 matrix);

    /**
     * Apply a Matrix4 to the plane. The matrix must be an affine, homogeneous transform.
     * If supplying an optionalNormalMatrix, it can be created like so:
     * Matrix3 optionalNormalMatrix = new Matrix3().getNormalMatrix( matrix );
     *
     * @param matrix - the Matrix4 to apply.
     * @param optionalNormalMatrix - (optional) pre-computed normal Matrix3 of the Matrix4 being applied.
     * @return
     */
    public native Plane applyMatrix4(Matrix4 matrix, Matrix3 optionalNormalMatrix);

    /**
     * Returns a new plane with the same normal and constant as this one.
     *
     * @return Plane
     */
    public native Plane clone();

    /**
     * Returns a Vector3 coplanar to the plane, by calculating the projection of the normal vector at the origin onto the plane.
     *
     * @return Vector3
     */
    public native Vector3 coplanarPoint();

    /**
     * Returns a Vector3 coplanar to the plane, by calculating the projection of the normal vector at the origin onto the plane.
     * optionalTarget if specified, the result will be copied into this Vector3, otherwise a new Vector3 will be created.
     *
     * @param optionalTarget
     * @return Plane
     */
    public native Vector3 coplanarPoint(Vector3 optionalTarget);

    /**
     * Copies the values of the passed plane's normal and constant properties to this plane.
     *
     * @param plane
     * @return Plane
     */
    public native Plane clone(Plane plane);

    /**
     * Returns the signed distance from the point to the plane.
     *
     * @param point
     * @return float
     */
    public native float distanceToPoint(Vector3 point);

    /**
     * Returns the signed distance from the sphere to the plane.
     *
     * @param sphere
     * @return float
     */
    public native float distanceToPoint(Sphere sphere);

    /**
     * Checks to see if two planes are equal (their normal and constant properties match).
     *
     * @param plane
     * @return boolean
     */
    public native boolean equals(Plane plane);

    /**
     * Returns the intersection point of the passed line and the plane. Returns undefined if the line does not intersect. Returns the
     * line's starting point if the line is coplanar with the plane.
     *
     * @param line - the Line3 to check for intersection.
     * @return Vector3
     */
    public native Vector3 intersectLine(Line3 line);

    /**
     * Returns the intersection point of the passed line and the plane. Returns undefined if the line does not intersect. Returns the
     * line's starting point if the line is coplanar with the plane.
     *
     * @param line           - the Line3 to check for intersection.
     * @param optionalTarget - the result will be copied into this Vector3, otherwise a new Vector3 will be created.
     * @return Vector3
     */
    public native Vector3 intersectLine(Line3 line, Vector3 optionalTarget);

    /**
     * Determines whether or not this plane intersects box.
     *
     * @param box - the Box3 to check for intersection.
     * @return boolean
     */
    public native boolean intersectsBox(Box3 box);

    /**
     * Tests whether a line segment intersects with (passes through) the plane.
     *
     * @param line - the Line3 to check for intersection.
     * @return
     */
    public native boolean intersectsLine(Line3 line);

    /**
     * Tests whether a line segment intersects with (passes through) the Sphere.
     *
     * @param sphere - the Sphere to check for intersection.
     * @return
     */
    public native boolean intersectsSphere(Sphere sphere);

    /**
     * Negates both the normal vector and the constant.
     *
     * @return Plane
     */
    public native Plane negate();

    /**
     * Normalizes the normal vector, and adjusts the constant value accordingly.
     *
     * @return Plane
     */
    public native Plane normalize();

    /**
     * Projects a point onto the plane.
     *
     * @param point - the Vector3 to project onto the plane.
     * @return Vector3
     */
    public native Vector3 projectPoint(Vector3 point);

    /**
     * Projects a point onto the plane.
     *
     * @param point          - the Vector3 to project onto the plane.
     * @param optionalTarget - the result will be copied into this Vector3, otherwise a new Vector3 will be created.
     * @return Vector3
     */
    public native Vector3 projectPoint(Vector3 point, Vector3 optionalTarget);

    /**
     * Sets the plane's normal and constant properties.
     *
     * @param normal   - a unit length Vector3 defining the normal of the plane.
     * @param constant - the signed distance from the origin to the plane. Default is 0.
     * @return Plane
     */
    public native Plane set(Vector3 normal, float constant);

    /**
     * Set the individual components that define the plane.
     *
     * @param x - x value of the unit length normal vector.
     * @param y - y value of the unit length normal vector.
     * @param z - z value of the unit length normal vector.
     * @param w - w value of the unit length normal vector.
     * @return Plane
     */
    public native Plane setComponents(float x, float y, float z, float w);

    /**
     * Defines the plane based on the 3 provided points. The winding order is assumed to be counter-clockwise, and determines the direction of the normal.
     *
     * @param a - first point on the plane.
     * @param b - second point on the plane.
     * @param c - third point on the plane.
     * @return Plane
     */
    public native Plane setFromCoplanarPoints(Vector3 a, Vector3 b, Vector3 c);

    /**
     * Sets the plane's properties as defined by a normal and an arbitrary coplanar point.
     *
     * @param normal - a unit length Vector3 defining the normal of the plane.
     * @param point  - Vector3
     * @return Plane
     */
    public native Plane setFromNormalAndCoplanarPoint(Vector3 normal, Vector3 point);

    /**
     * Translates the plane by the distance defined by the offset vector. Note that this only affects the plane constant
     * and will not affect the normal vector.
     *
     * @param offset - the amount to move the plane by.
     * @return Plane
     */
    public native Plane translate (Vector3 offset);
}

File: core/src/main/resources/org/treblereel/gwt/three4g/math/Plane.js
import { Matrix3 } from './Matrix3.js';
import { Matrix4 } from './Matrix4.js';
import { Vector3 } from './Vector3.js';

//closure compiler
import { Box3Interface } from '../closure/math/Box3Interface.js';
import { PlaneInterface } from '../closure/math/PlaneInterface.js';
import { SphereInterface } from '../closure/math/SphereInterface.js';

const _vector1 = /*@__PURE__*/ new Vector3();
const _vector2 = /*@__PURE__*/ new Vector3();
const _normalMatrix = /*@__PURE__*/ new Matrix3();

/**
 * @implements { PlaneInterface }
 */
class Plane {

	/**
	 * @param {Vector3=} normal 
	 * @param {number=} constant 
	 */
	constructor( normal = new Vector3( 1, 0, 0 ), constant = 0 ) {

		// normal is assumed to be normalized

		/** @type {Vector3} */
		this.normal = normal;
		/** @type {number} */
		this.constant = constant;

	}

	/**
	 * @param {Vector3} normal 
	 * @param {number} constant
	 * @return {Plane} 
	 */
	set( normal, constant ) {

		this.normal.copy( normal );
		this.constant = constant;

		return this;

	}

	/**
	 * @param {number} x 
	 * @param {number} y 
	 * @param {number} z 
	 * @param {number} w 
	 * @return {Plane} 
	 */
	setComponents( x, y, z, w ) {

		this.normal.set( x, y, z );
		this.constant = w;

		return this;

	}

	/**
	 * @param {Vector3} normal 
	 * @param {Vector3} point
	 * @return {Plane} 
	 */
	setFromNormalAndCoplanarPoint( normal, point ) {

		this.normal.copy( normal );
		this.constant = - point.dot( this.normal );

		return this;

	}

	/**
	* @param {Vector3} a 
	* @param {Vector3} b
	* @param {Vector3} c
	* @return {Plane} 
	*/
	setFromCoplanarPoints( a, b, c ) {

		const normal = _vector1.subVectors( c, b ).cross( _vector2.subVectors( a, b ) ).normalize();

		// Q: should an error be thrown if normal is zero (e.g. degenerate plane)?

		this.setFromNormalAndCoplanarPoint( normal, a );

		return this;

	}

	/**
	* @param {Plane} plane
	* @return {Plane} 
	*/
	copy( plane ) {

		this.normal.copy( plane.normal );
		this.constant = plane.constant;

		return this;

	}

	/**
	* @return {Plane} 
	*/
	normalize() {

		// Note: will lead to a divide by zero if the plane is invalid.

		const inverseNormalLength = 1.0 / this.normal.length();
		this.normal.multiplyScalar( inverseNormalLength );
		this.constant *= inverseNormalLength;

		return this;

	}

	/**
	* @return {Plane} 
	*/
	negate() {

		this.constant *= - 1;
		this.normal.negate();

		return this;

	}

	/**
	* @param {Vector3} point
	* @return {number} 
	*/
	distanceToPoint( point ) {

		return this.normal.dot( point ) + this.constant;

	}

	/**
	 * @param { SphereInterface } sphere
	* @return {number} 
	*/
	distanceToSphere( sphere ) {

		return this.distanceToPoint( sphere.center ) - sphere.radius;

	}

	/**
	* @param {Vector3} point
	* @param {Vector3=} target
	* @return {Vector3} 
	*/
	projectPoint( point, target ) {

		if ( target === undefined ) {

			console.warn( 'THREE.Plane: .projectPoint() target is now required' );
			target = new Vector3();

		}

		return target.copy( this.normal ).multiplyScalar( - this.distanceToPoint( point ) ).add( point );

	}

	/**
	 * @param { { delta : function(Vector3) : Vector3, start : Vector3 } } line
	 * @param {Vector3} target
	 * @return {Vector3|undefined} 
	 */
	intersectLine( line, target ) {

		if ( target === undefined ) {

			console.warn( 'THREE.Plane: .intersectLine() target is now required' );
			target = new Vector3();

		}

		const direction = line.delta( _vector1 );

		const denominator = this.normal.dot( direction );

		if ( denominator === 0 ) {

			// line is coplanar, return origin
			if ( this.distanceToPoint( line.start ) === 0 ) {

				return target.copy( line.start );

			}

			// Unsure if this is the correct method to handle this case.
			return undefined;

		}

		const t = - ( line.start.dot( this.normal ) + this.constant ) / denominator;

		if ( t < 0 || t > 1 ) {

			return undefined;

		}

		return target.copy( direction ).multiplyScalar( t ).add( line.start );

	}

	/**
	 * @param {{start : Vector3, end: Vector3}} line
	* @return {boolean} 
	*/
	intersectsLine( line ) {

		// Note: this tests if a line intersects the plane, not whether it (or its end-points) are coplanar with it.

		const startSign = this.distanceToPoint( line.start );
		const endSign = this.distanceToPoint( line.end );

		return ( startSign < 0 && endSign > 0 ) || ( endSign < 0 && startSign > 0 );

	}

	/**
	 * @param { Box3Interface } box
	* @return {boolean} 
	*/
	intersectsBox( box ) {

		return box.intersectsPlane( this );

	}

	/**
	 * @param { SphereInterface } sphere
	* @return {boolean} 
	*/
	intersectsSphere( sphere ) {

		return sphere.intersectsPlane( this );

	}

	/**
	* @param {Vector3} target 
	* @return {Vector3} 
	*/
	coplanarPoint( target ) {

		if ( target === undefined ) {

			console.warn( 'THREE.Plane: .coplanarPoint() target is now required' );
			target = new Vector3();

		}

		return target.copy( this.normal ).multiplyScalar( - this.constant );

	}

	/**
	 * 
	 * @suppress {checkTypes} 
	 * @param {Matrix4} matrix
	 * @param {Matrix3} optionalNormalMatrix
	 * @return {Plane} 
	*/
	applyMatrix4( matrix, optionalNormalMatrix ) {

		const normalMatrix = optionalNormalMatrix || _normalMatrix.getNormalMatrix( matrix );

		const referencePoint = this.coplanarPoint( _vector1 ).applyMatrix4( matrix );

		const normal = this.normal.applyMatrix3( normalMatrix ).normalize();

		this.constant = - referencePoint.dot( normal );

		return this;

	}

	/**
	 * @param {Vector3} offset
	* @return {Plane} 
	*/
	translate( offset ) {

		this.constant -= offset.dot( this.normal );

		return this;

	}

	/**
	 * @param {Plane} plane
	 * @return {boolean} 
	 */
	equals( plane ) {

		return plane.normal.equals( this.normal ) && ( plane.constant === this.constant );

	}

	/**
	 * @return {Plane} 
	 */
	clone() {

		return new this.constructor().copy( this );

	}

}

Plane.prototype.isPlane = true;

export { Plane };
